### PR TITLE
feat: [CDS-52288]: Auto Reject previous approvals

### PR DESCRIPTION
**Pull request by Y0709**
We have added support for a new feature to reject previous waiting executions on the same approval step for same pipeline on approval of the latest execution.
We have added a boolean field in yaml to check whether this feature is configured or not and
We have added 2 new fields in harness approval instances to check if auto rejection is configured or not and a approval key to check which all approvals are to be rejected. 
The value for this approval key for is (stageIdentifier)#(stepIdentifier)
At some point we can add another field in yaml to provide this approval key.

We have pr tested this feature in pr env as well.
https://docs.google.com/spreadsheets/d/1CpkYdw_pr1QK9MfzKsWon422wUZ8X72JN3V891EnE_w/edit#gid=214881638

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/45666)

### CODE REVIEW

**Comment by Y0709** [comment]
trigger messagecheck

**Comment by Y0709** [comment]
trigger ut4

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/step/ApprovalInstanceServiceImpl.java [hunk 1]
```diff
@@ -223,6 +226,70 @@ public HarnessApprovalInstance addHarnessApprovalActivity(@NotNull String approv
...
+  @Override
+  public List<String> findAllPreviousWaitingApprovals(String accountId, String orgId, String projectId,
+      @NotEmpty String pipelineId, @Valid ApprovalStatus approvalStatus, String approvalKey, Ambiance ambiance) {
+    Criteria criteria = Criteria.where("pipelineIdentifier").is(pipelineId);
+    criteria.and("isAutoRejectEnabled").is(true);
+
+    if (!isNull(approvalStatus)) {
+      criteria.and(ApprovalInstanceKeys.status).is(approvalStatus);
+    }
+    if (!isNull(accountId)) {
```
Comment (puthrayaharness) [inline pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/step/ApprovalInstanceServiceImpl.java:255]: `accountId` should always be used with all queries. Please add a notNull check if you expect `accountId` to be coming as `null`.

**Review by puthrayaharness — CHANGES_REQUESTED** [review]

File: pipeline-service/modules/orchestration-steps/contracts/src/main/java/io/harness/steps/approval/step/harness/beans/HarnessApprovalInstanceDetailsDTO.java [hunk 2]
```diff
@@ -35,4 +35,6 @@ public class HarnessApprovalInstanceDetailsDTO implements ApprovalInstanceDetail
   List<HarnessApprovalActivity> approvalActivities;
   List<ApproverInputInfoDTO> approverInputs;
   List<ApprovalUserGroupDTO> validatedApprovalUserGroups;
+  String approvalKey;
```
Comment (naman-goenka) [inline pipeline-service/modules/orchestration-steps/contracts/src/main/java/io/harness/steps/approval/step/harness/beans/HarnessApprovalInstanceDetailsDTO.java:38]: We can remove approvalKey from DTO and mapper in first phase

Comment (Y0709) [reply pipeline-service/modules/orchestration-steps/contracts/src/main/java/io/harness/steps/approval/step/harness/beans/HarnessApprovalInstanceDetailsDTO.java]: done

File: pipeline-service/service/src/main/java/io/harness/pms/approval/ApprovalResourceServiceImpl.java [hunk 3]
```diff
@@ -85,15 +87,36 @@ public ApprovalInstanceResponseDTO get(String approvalInstanceId) {
...
+        accountId, orgId, projectId, pipelineId, ApprovalStatus.WAITING, approvalKey, ambiance);
+    rejectedApprovalIds.forEach(id -> {
+      boolean unauthorized = !getHarnessApprovalInstanceAuthorization(id, true).isAuthorized();
+      approvalInstanceService.rejectPreviousExecutions(id, getEmbeddedUser(), unauthorized, ambiance);
+    });
+  }
+
   private EmbeddedUser getEmbeddedUser() {
     Principal principal = currentUserHelper.getPrincipalFromSecurityContext();
     if (!USER.equals(principal.getType())) {
```
Comment (naman-goenka) [inline pipeline-service/service/src/main/java/io/harness/pms/approval/ApprovalResourceServiceImpl.java:122]: Do we need to support api-key here?

Comment (Y0709) [reply pipeline-service/service/src/main/java/io/harness/pms/approval/ApprovalResourceServiceImpl.java:122]: Not sure about this but this is existing behaviour only

**Review by naman-goenka — APPROVED** [review]

File: pipeline-service/service/src/main/java/io/harness/pms/approval/ApprovalResourceServiceImpl.java (hunk 3 above)
Comment (naman-goenka) [reply pipeline-service/service/src/main/java/io/harness/pms/approval/ApprovalResourceServiceImpl.java:136]: ok, resolving this for now

**Pull request merged**